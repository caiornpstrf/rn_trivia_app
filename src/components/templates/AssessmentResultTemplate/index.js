/**
 * @summary Summary of the result of an Assessment
 * @author Caio Reis <caio.oliveira.reis@gmail.com>
 *
 * Created at     : 2021-04-16 03:45:13
 * Last modified  : 2021-04-16 03:46:11
 */

import React from 'react';
import {StyleSheet, View} from 'react-native';

import {Button, Title, Description} from '_components/atoms';
import {Colors} from '_assets/styles';
import {AssessmentStrings} from '_assets/constants';
import {string} from 'prop-types';
import {func} from 'prop-types';
import {object} from 'prop-types';

const placeListOfVariablesOnString = (label = '', list = []) => {
  let newLabel = label;
  for (let i = 0; i < list.length; i++) {
    newLabel = newLabel.replace(`{${i}}`, list[i]);
  }
  return newLabel;
};

const AssessmentResultTemplate = ({buttonLabel, onPress, resultsMapping}) => {
  const {result, resultDetails} = AssessmentStrings;
  return (
    <View style={styles.container}>
      <View style={styles.textContainer}>
        <Title size="big" style={styles.text}>
          {placeListOfVariablesOnString(result, resultsMapping.summary)}
        </Title>
        <Title size="medium" style={styles.text}>
          {resultDetails.title}
        </Title>
        <Description style={styles.text}>
          {placeListOfVariablesOnString(
            resultDetails.easy,
            resultsMapping.easy,
          )}
          {'\n'}
          {placeListOfVariablesOnString(
            resultDetails.medium,
            resultsMapping.medium,
          )}
          {'\n'}
          {placeListOfVariablesOnString(
            resultDetails.hard,
            resultsMapping.hard,
          )}
        </Description>
      </View>
      <Button label={buttonLabel} onPress={onPress} type="answer" />
    </View>
  );
};

AssessmentResultTemplate.propTypes = {
  buttonLabel: string,
  onPress: func,
  resultsMapping: object,
};

AssessmentResultTemplate.defaultProps = {
  buttonLabel: '',
  onPress: () => console.log('Pressed AssessmentResultTemplate button'),
  resultsMapping: {summary: [0, 0], easy: [0, 0], medium: [0, 0], hard: [0, 0]},
};

export default AssessmentResultTemplate;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    paddingHorizontal: 10,
    paddingVertical: 20,
    backgroundColor: Colors.background,
  },
  textContainer: {
    flex: 1,
    backgroundColor: Colors.background,
    paddingHorizontal: 20,
  },
  text: {
    marginBottom: 25,
  },
});
