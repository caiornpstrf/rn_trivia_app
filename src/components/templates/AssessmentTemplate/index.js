/**
 * @summary Displays every info of a Question, allows for answer selection & displays the result
 * @author Caio Reis <caio.oliveira.reis@gmail.com>
 *
 * Created at     : 2021-04-16 03:46:18
 * Last modified  : 2021-04-16 03:47:37
 */

import React from 'react';
import {StyleSheet, View, Modal, Pressable} from 'react-native';
import {array, func, string, bool} from 'prop-types';

import {Button, IfElse} from '_components/atoms';
import {QuestionDetails, QuestionResult} from '_components/molecules';
import {AnswersList} from '_components/organisms';
import {Colors} from '_assets/styles';

const AssessmentTemplate = ({
  title,
  question,
  difficulty,
  answers,
  onItemSelected,
  displayNextButton,
  nextButtonLabel,
  onPressNextButton,
  modalVisible,
  onRequestClose,
  correctLabel,
  incorrectLabel,
  isCorrect,
}) => {
  return (
    <View style={styles.container}>
      <QuestionDetails
        title={title}
        question={question}
        difficulty={difficulty}
      />
      <AnswersList data={answers} onItemSelected={onItemSelected} />
      <IfElse
        condition={displayNextButton}
        onTrue={
          <View style={styles.buttonContainer}>
            <Button
              type="answer"
              label={nextButtonLabel}
              onPress={onPressNextButton}
            />
          </View>
        }
      />
      <Modal
        animationType="fade"
        visible={modalVisible}
        onRequestClose={onRequestClose}
        transparent>
        <Pressable style={styles.modal} onPress={onRequestClose}>
          <QuestionResult
            correctLabel={correctLabel}
            incorrectLabel={incorrectLabel}
            isCorrect={isCorrect}
          />
        </Pressable>
      </Modal>
    </View>
  );
};

AssessmentTemplate.propTypes = {
  title: string.isRequired,
  question: string.isRequired,
  difficulty: string.isRequired,
  answers: array.isRequired,
  onItemSelected: func.isRequired,
  displayNextButton: bool.isRequired,
  nextButtonLabel: string.isRequired,
  onPressNextButton: func.isRequired,
  modalVisible: bool,
  onRequestClose: func,
  correctLabel: string,
  incorrectLabel: string,
  isCorrect: bool,
};

AssessmentTemplate.defaultProps = {
  modalVisible: false,
  onRequestClose: () => console.log('Requested Modal to close'),
  correctLabel: '',
  incorrectLabel: '',
  isCorrect: false,
};

export default AssessmentTemplate;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    paddingHorizontal: 10,
    paddingVertical: 20,
    backgroundColor: Colors.background,
  },
  buttonContainer: {
    backgroundColor: Colors.background,
    paddingHorizontal: '15%',
    paddingVertical: '5%',
    justifyContent: 'center',
    alignContent: 'center',
  },
  modal: {
    flex: 1,
    paddingHorizontal: '20%',
    justifyContent: 'center',
    alignContent: 'center',
  },
});
