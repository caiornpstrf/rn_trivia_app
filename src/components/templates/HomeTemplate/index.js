import React from 'react';
import {StyleSheet, View} from 'react-native';
import {array, func, string} from 'prop-types';

import {Title} from '_components/atoms';
import {CategoryList} from '_components/organisms';
import {Colors} from '_assets/styles';

const HomeTemplate = ({title, categories, onItemSelected}) => {
  return (
    <View style={styles.container}>
      <Title size="big" style={styles.title}>
        {title}
      </Title>
      <CategoryList data={categories} onItemSelected={onItemSelected} />
    </View>
  );
};

HomeTemplate.propTypes = {
  title: string.isRequired,
  categories: array.isRequired,
  onItemSelected: func.isRequired,
};

export default HomeTemplate;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    paddingHorizontal: 20,
    paddingVertical: 20,
    backgroundColor: Colors.background,
  },
  title: {
    marginBottom: 40,
  },
});
