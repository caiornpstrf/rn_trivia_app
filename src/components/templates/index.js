export {default as AssessmentTemplate} from './AssessmentTemplate';
export {default as AssessmentResultTemplate} from './AssessmentResultTemplate';
export {default as HomeTemplate} from './HomeTemplate';
