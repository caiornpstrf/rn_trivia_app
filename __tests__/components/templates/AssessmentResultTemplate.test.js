import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {AssessmentResultTemplate} from '_components/templates';

const defaultProps = {
  buttonLabel: 'Screen button',
  resultsMapping: {
    summary: [8, 10],
    easy: [2, 2],
    medium: [2, 3],
    hard: [5, 5],
  },
  onPress: jest.fn(),
};

it('renders correctly', () => {
  const {toJSON} = render(<AssessmentResultTemplate {...defaultProps} />);
  expect(toJSON()).toMatchSnapshot();
});

it('presses the main button successfully', () => {
  const {getByText} = render(<AssessmentResultTemplate {...defaultProps} />);
  fireEvent.press(getByText(defaultProps.buttonLabel));
  expect(defaultProps.onPress).toBeCalled();
});
