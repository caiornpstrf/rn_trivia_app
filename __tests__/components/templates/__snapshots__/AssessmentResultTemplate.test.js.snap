// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingVertical": 20,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "paddingHorizontal": 20,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#343D58",
            "fontWeight": "bold",
          },
          Object {
            "marginBottom": 25,
          },
          Object {
            "fontSize": 42,
          },
        ]
      }
    >
      Overall performance: 8 of 10 correct answers
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#343D58",
            "fontWeight": "bold",
          },
          Object {
            "marginBottom": 25,
          },
          Object {
            "fontSize": 35,
          },
        ]
      }
    >
      Details:
    </Text>
    <Text
      style={
        Object {
          "color": "#171717",
          "fontSize": 28,
        }
      }
    >
      • Easy: 2 of 2
      

      • Medium: 2 of 3
      

      • Hard: 5 of 5
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#4D8AF0",
          "borderRadius": 10,
          "elevation": 4,
          "height": 45,
          "justifyContent": "center",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.23,
          "shadowRadius": 2.62,
        },
        undefined,
        false,
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 33,
          "fontWeight": "bold",
        }
      }
    >
      Screen button
    </Text>
  </View>
</View>
`;
